keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/MultinomialVertexTest.java

```diff
@@ -47,221 +46,436 @@
+        IntegerVertex n = ConstantVertex.of(10);
+        DoubleVertex p = ConstantVertex.of(DoubleTensor.create(0.2, 0.8));
+        MultinomialVertex multinomial = new MultinomialVertex(n, p);
+        multinomial.setValidationEnabled(true);
+        multinomial.logProb(IntegerTensor.create(5, 5, 3, 7).reshape(2, 2));
+    }
+
+    @Test
     public void itThrowsIfTheLogProbStateContainsNegativeNumbers() {
-        IntegerTensor n = IntegerTensor.scalar(10).reshape(1, 1);
-        DoubleTensor p = DoubleTensor.create(0.2, 0.8).reshape(2, 1);
-        DiscreteDistribution multinomial = Multinomial.withParameters(n, p);
+        thrown.expect(IllegalArgumentException.class);
+        thrown.expectMessage("x must be >= 0 and <= n");
+
+        IntegerTensor n = IntegerTensor.scalar(10);
+        DoubleTensor p = DoubleTensor.create(0.2, 0.8);
+        MultinomialVertex multinomial = new MultinomialVertex(n, p);
+        multinomial.setValidationEnabled(true);
         multinomial.logProb(IntegerTensor.create(-1, 11).transpose());
     }
 
-    @Test(expected = IllegalArgumentException.class)
+    @Test
     public void itThrowsIfTheLogProbStateContainsNumbersGreaterThanN() {
-        IntegerTensor n = IntegerTensor.scalar(10).reshape(1, 1);
-        DoubleTensor p = DoubleTensor.create(0.2, 0.3, 0.5).reshape(3, 1);
-        DiscreteDistribution multinomial = Multinomial.withParameters(n, p);
+        thrown.expect(IllegalArgumentException.class);
+        thrown.expectMessage("x must be >= 0 and <= n");
+
+        IntegerTensor n = IntegerTensor.scalar(10);
+        DoubleTensor p = DoubleTensor.create(0.2, 0.3, 0.5);
+        MultinomialVertex multinomial = new MultinomialVertex(n, p);
+        multinomial.setValidationEnabled(true);
         int[] state = new int[]{Integer.MAX_VALUE, Integer.MAX_VALUE, 12};
         assertThat(state[0] + state[1] + state[2], equalTo(10));
         multinomial.logProb(IntegerTensor.create(state).transpose());
     }
 
     @Test
-    public void itWorksWithScalars() {
+    public void itWorksWithScalarTrialCountN() {
+        int n = 4;
+        DoubleTensor p = DoubleTensor.create(new double[]{0.2, 0.3, 0.5}, 3);
+
+        MultinomialVertex multinomial = new MultinomialVertex(n, p);
+        IntegerTensor sample = multinomial.sample();
+
+        assertThat(sample, hasShape(3));
+        assertThat(sample, allValues(both(greaterThan(-1)).and(lessThan(n))));
+        assertThat(sample.sum(), equalTo(n));
+    }
+
+    @Test
+    public void itWorksWithAlmostOneProbability() {
         int n = 100;
-        DoubleTensor p = DoubleTensor.create(new double[]{0.01, 0.09, 0.9}, 3, 1);
-        MultinomialVertex multinomial = new MultinomialVertex(n, ConstantVertex.of(p));
-        IntegerTensor samples = multinomial.sample(KeanuRandom.getDefaultRandom());
-        assertThat(samples, hasShape(3, 1));
+        DoubleTensor p = DoubleTensor.create(0.1, new long[]{10});
```

Comment:
this is meant to be almost one? This `p` is exactly `1.0`

Reply:
The sum of those 0.1s is actually `0.9999999999999999` due to floating point summation. This was causing this exact case to throw an exception before.

Reply:
you should express that in the test - `assertThat(p.sum(), not(equalTo(1.0))`